src/controller.rs

```diff
@@ -1,3 +1,118 @@
+//! The _requeue_ can happen immediately or after a custom duration.
+//! The latter can be useful for operations that are expected to take a while (e.g. creating a resource which is under control by a different Operator).
+//!
+//! This module has a [`Controller`] which is the main object that drives the reconciliation loop.
+//! Custom logic needs to be plugged in by implementing a [`ControllerStrategy`] trait.
+//! The `Controller` will call `init_reconcile_state(...)` on the `ControllerStrategy` for every reconciliation that is triggered.
+//! This method returns a `ReconciliationState` trait which can contain arbitrary state about this
+//! specific reconciliation run.
+//!
+//! Once the state object has been created its `reconcile` method is called.
+//! The `reconcile` method is where the business logic for each operator gets called.
+//! `ReconcileFunctionAction`s can be used to chain multiple calls to separate methods.
+//! By using the `then` method on `ReconcileFunctionAction` we can automatically abort and/or requeue
+//! in case of an `Error` or a `ReconciliationFunctionAction::Requeue(duration)`.
+//!
+//! See the example below for how to use this abstraction in a real-world operator.
+//!
+//! # Example
+//!
+//! ```no_run
+//! use kube::Api;
+//! use k8s_openapi::api::core::v1::Pod;
+//! use stackable_operator::client;
+//! use stackable_operator::controller::{Controller, ControllerStrategy, ReconciliationState};
+//! use stackable_operator::reconcile::{ReconciliationContext, ReconcileFunctionAction, ReconcileResult};
+//! use std::pin::Pin;
+//! use std::future::Future;
+//! use std::time::Duration;
+//!
+//! struct FooStrategy {
+//! }
+//!
+//! struct FooState {
+//!     my_state: i32
+//! }
+//!
+//! type FooReconcileResult = ReconcileResult<String>;
+//!
+//! impl FooState {
+//!     async fn test1(&mut self) -> FooReconcileResult {
+//!         self.my_state = 123;
+//!         println!("My reconciliation logic part 1");
+//!         Ok(ReconcileFunctionAction::Continue)
+//!     }
+//!
+//!     async fn test2(&self) -> FooReconcileResult {
+//! println!("My reconciliation logic part 2");
+//!         if self.my_state > 100 {
+//!             return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+//!         }   
+//!         Ok(ReconcileFunctionAction::Continue)
+//!     }
+//!
+//!     async fn test3(&self) -> FooReconcileResult {
+//! println!("My reconciliation logic part 3");
+//!         if self.my_state > 100 {
+//!             return Ok(ReconcileFunctionAction::Done);
+//!         }   
+//!         Ok(ReconcileFunctionAction::Continue)
+//!     }
+//! }
+//!
+//! impl ReconciliationState for FooState {
+//!     type Error = String;
+//!
+//!     fn reconcile(
+//!         &mut self,
+//!     ) -> Pin<Box<dyn Future<Output = Result<ReconcileFunctionAction, Self::Error>> + Send + '_>>
+//!     {
+//!         Box::pin(async move {
+//!             self.test1()
```

Comment:
It looks a bit weird the way that subsequent calls are at a different level than the first one due to them being chained after that one.
But that really is just an observation, no need to do anything about this!

My first instinct was to give the ReconcileFunctionAction an associated function to execute all of them, but most probably that would reintroduce the exact problem that you worked hard to avoid, so disregard this ;)

Reply:
I understand about the initial issue but I don't understand your proposed solution but I have a feeling that your hunch might be correct :) Lots of hunches though!

Reply:
As I said, can be ignored :)